network/netcore/src/multiplexing/yamux.rs

```diff
@@ -76,152 +67,127 @@ where
 {
-    type Item = io::Result<Compat01As03<yamux::StreamHandle<Compat<TSocket>>>>;
-
-    fn poll_next(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Option<Self::Item>> {
-        match Pin::new(&mut self.inner).poll_next(context) {
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(None) => Poll::Ready(None),
-            Poll::Ready(Some(Err(e))) => {
-                Poll::Ready(Some(Err(io::Error::new(io::ErrorKind::Other, e))))
+    type Substream = StreamHandle;
+    type Control = YamuxControl;
+    type Listener = BoxStream<'static, io::Result<StreamHandle>>;
+
+    async fn start(self) -> (Self::Listener, Self::Control) {
+        let control = self.connection.control();
+        let mut connection = self.connection;
+        let (mut tx, rx) = mpsc::unbounded();
+        // Our library wraps calls to `next_stream` in a future and runs the future on this static
+        // tokio runtime. That way, users of our yamux library don't need to worry about the
+        // unintuitive need for polling for new inbound substreams even if they only need to open
+        // outbound substreams or perform IO on existing susbstreams.
+        let f = async move {
+            loop {
+                let maybe_substream = connection.next_stream().await.transpose();
+                match maybe_substream {
+                    None | Some(Err(_)) => {
+                        // Ignore error.
+                        let _ = tx
+                            .send(Err(io::Error::new(
+                                io::ErrorKind::BrokenPipe,
+                                "Connection closed",
+                            ))
+                                as io::Result<Self::Substream>)
+                            .await;
+                        return;
+                    }
+                    Some(Ok(substream)) => {
+                        if tx.send(Ok(substream)).await.is_err() {
+                            // Failed to notify about new inbound substream.
+                            return;
+                        }
+                    }
+                }
             }
-            Poll::Ready(Some(Ok(substream))) => Poll::Ready(Some(Ok(Compat01As03::new(substream)))),
-        }
+        };
+        // Run the IO future on the executor which drives this connection.
+        tokio::spawn(f);
+        (rx.boxed(), YamuxControl { inner: control })
     }
 }
 
 #[cfg(test)]
 mod test {
     use crate::multiplexing::{
         yamux::{Mode, Yamux},
-        StreamMultiplexer,
+        Control, StreamMultiplexer,
     };
     use futures::{
         executor::block_on,
         future::join,
         io::{AsyncReadExt, AsyncWriteExt},
         stream::StreamExt,
     };
+    use libra_logger::prelude::*;
     use memsocket::MemorySocket;
     use std::io;
 
-    #[test]
-    fn substream_within_substream() -> io::Result<()> {
+    #[tokio::test(core_threads = 2)]
```

Comment:
`core_threads=2` was needed since we had a `block_on` in the body of the test function which is itself async. Changing the `block_on` to `await` removed the need to have more than 1 thread. Thanks for pointing out though.